posts/inside-rust/2022-11-17-async-fn-in-trait-nightly.md

```diff
@@ -0,0 +1,290 @@
+
+```rust
+impl Database for MyDb {
+    type FetchData<'a> = /* what type goes here??? */;
+    fn fetch_data(&self) -> FetchData<'a> { async move { ... } }
+}
+```
+
+Since you can't name the type constructed by an async block, the only option is to use an opaque type (the `impl Future` we saw earlier). But those are not supported in associated types![^tait]
+
+[^tait]: This feature is called ["type alias impl trait"](https://rust-lang.github.io/rfcs/2515-type_alias_impl_trait.html).
+
+### Workarounds available in the stable compiler
+
+So to write an `async fn` in a trait we need a concrete type to specify in our impl, if not our trait. There are a couple ways of achieving this today.
+
+#### Runtime type erasure
+
+First, we can avoid writing the future type by erasing it with `dyn`. Taking our example from above, you would write your trait like this:
+
+```rust
+trait Database {
+    fn fetch_data(&self)
+    -> Pin<Box<dyn Future<Output = String> + Send + '_>>;
+}
+```
+
+This is significantly more verbose, but it achieves the goal of combining async with traits. What's more, the [async-trait] proc macro crate rewrites your code for you, allowing you to simply write
+
+```rust
+#[async_trait]
+trait Database {
+    async fn fetch_data(&self) -> String;
+}
+
+#[async_trait]
+impl Database for MyDb {
+    async fn fetch_data(&self) -> String { ... }
+}
+```
+
+This is an excellent solution for the people who can use it!
+
+Unfortunately, not everyone can. You can't use `Box` in no_std contexts. Dynamic dispatch and allocation come with overhead that can be [crippling][barbara-benchmark] to highly performance-sensitive code. Finally, it bakes a lot of assumptions into the trait itself: allocation with `Box`, dynamic dispatch, and the `Send`-ness of the futures. This makes it unsuitable for many libraries.
+
+Besides, users [expect][alan-async-traits] to be able to write `async fn` in traits, and the experience of adding an external crate dependency is a papercut that gives async Rust a reputation for being difficult to use.
+
+#### Manual `poll` implementations
+
+Traits that need to work with zero overhead or in no_std contexts have another option: they can take the concept of polling from the [`Future` trait](https://doc.rust-lang.org/stable/std/future/trait.Future.html) and build it directly into their interface. The `Future::poll` method returns `Poll::Ready(Output)` if the future is complete and `Poll::Pending` if the future is waiting on some other event.
+
+You can see this pattern, for example, in the Stream[^stream-futures] trait. The signature of `Stream::poll_next` is a cross between `Future::poll` and `Iterator::next`.
+
+[^stream-futures]: The [`Stream` trait](https://docs.rs/futures/latest/futures/stream/trait.Stream.html) is part of the `futures` crate, not the standard library.
```

Comment:
AsyncIterator is in std, so it might be less confusing to use that in the examples? Or add a note about this to this footnote